src/Data/Tree/AVL/Internal.hs

```diff
@@ -0,0 +1,274 @@
+
+{-# language FunctionalDependencies #-}
+{-# language LambdaCase #-}
+{-# language MultiParamTypeClasses #-}
+{-# language TemplateHaskell #-}
+{-# language NamedFieldPuns #-}
+{-# language PatternSynonyms #-}
+{-# language StrictData #-}
+{-# language FlexibleInstances #-}
+{-# language FlexibleContexts #-}
+{-# language ScopedTypeVariables #-}
+{-# language RankNTypes #-}
+{-# language DeriveFunctor #-}
+{-# language DeriveGeneric              #-}
+{-# language StandaloneDeriving #-}
+{-# language DeriveAnyClass #-}
+{-# language UndecidableInstances #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+module Data.Tree.AVL.Internal where
+
+import Control.Lens ( (&), (^.), (.~), (^?)
+                    , to
+                    , makeLenses, makePrisms
+                    , Getter, Setter', Lens'
+                    )
+import Data.Binary
+import Data.Default (Default(..))
+import Data.Fix     (Fix(..))
+
+import GHC.Generics hiding (to)
+
+
+-- | "Tilt" is a difference in branch heights.
+--   We have only +2/-2 as our limits for the rebalance, lets make enum.
+--   1) This will be compiled to number comparison;
+--   2) This will die fast in case some part of the algorithm tries to inc R2.
+data Tilt
+    = L2  -- UnbalancedLeft
+    | L1  -- BalancedLeft
+    | M   -- Balanced
+    | R1  -- BalancedRight
+    | R2  -- UnbalancedRight
+    deriving (Eq, Ord, Enum, Show, Generic, Binary)
+
+-- | Representation of AVL+ tree with data in leaves.
+
+type Map h k v = Fix (MapLayer h k v)
```

Comment:
The reason was the hashing. Since the implementer of `Hash` interface for the library is the client code, it should be prevented from doing things like going inside nodes and must follow the shape of the tree. See `rehash` - it also effectively prevents `hashOf` from depending on previous hash.

Reply:
About differences from paper. This is not the paper we used with @vlad107, actually. I followed more IOHK's scala implementation (and their WP(?)), which stores `centerKey` instead `minKey`. However, you cannot get `centerKey` for the branch w/o obtaining `minKey` for the right node.

And they seemed to prove that you must include all the meaningful fields inside the hash (and proof) - balance parameter, center key and such. If our impl contains at least these - we're safe.

The `centerKey` is used for navigation. I don't know the way to navigate inside the tree without it.
And without the `minKey` geting `centerKey` w/o going all the way down the left branch is impossible.

Later we decided with @vlad107 that nodes should have not only `nextKey`, but a `prevKey` as well - because when you `delete`, the `nextKey`-chain will become inconsistent _unless you traverse to previous key manually and fix it_. This decision was made before I put project on zipper rails, so "going to the previous key" was troublesome.

Now I can fix it in two ways:
1) go to the previous key by the means of zipper, set `nextKey` there - without storing `prevKey`;
2) visit (using zipper) and `mark` both previous and next key, so they both get into the proof, and the consumer of it will do the same to check that node [does not] exists between them.

> AVL+ proofs do not contain keys of internal nodes

AVL+ tree doesn't have any keys, belonging to the internal nodes. The keys there are for navigation.

I don't know how to remove them from proof and be able to prove that we did `delete`, 7 rebalances and ended with _this_ hash. And the IOHK impl/WP does include `centerKey` in both proof and hash. If short: the proof must include at least the fields the hash does. And the hash should notice the changes in the topology and need of rebalance.

Another problem about excluding fields from the proof is that some operations require the traversed path in the tree to fork and some point. This is hard to represent, unless we use tree slice as proof.

And the tree slice adds us an ability to arbitrarily choose at what level we need proofs: we can prove each transaction OR the whole block of them (and win some space due to deduplication of root nodes).

Reply:
I can remove `centerKey` and go into `right` branch to get myself a `minKey` to use it as former. This will shorten a proof, but the traversal time will increase. So I'd rather left `centerKey` be until we have benchmarks.